Resources/js/router.ts

```diff
@@ -0,0 +1,350 @@
+      if (variants[i] in this.routes_) {
+        return this.routes_[variants[i]];
+      }
+    }
+
+    throw new Error('The route "' + name + '" does not exist.');
+  }
+
+  /**
+   * Generates the URL for a route.
+   */
+  generate(name: string, opt_params?: RouteParams, absolute?: boolean): string {
+    let route = (this.getRoute(name));
+    let params = opt_params || {};
+    let unusedParams = Object.assign({}, params);
+    let url = '';
+    let optional = true;
+    let host = '';
+    let port = (typeof this.getPort() == 'undefined' || this.getPort() === null) ? '' : this.getPort();
+
+    route.tokens.forEach((token) => {
+      if ('text' === token[0] && typeof token[1] === 'string') {
+        url = Router.encodePathComponent(token[1]) + url;
+        optional = false;
+
+        return;
+      }
+
+      if ('variable' === token[0]) {
+        let hasDefault = route.defaults && !Array.isArray(route.defaults) && typeof token[3] === 'string' && (token[3] in route.defaults);
+        if (false === optional || !hasDefault || ((typeof token[3] === 'string' && token[3] in params) && !Array.isArray(route.defaults) && params[token[3]] != route.defaults[token[3]])) {
+          let value;
+
+          if (typeof token[3] === 'string' && token[3] in params) {
+            value = params[token[3]];
+            delete unusedParams[token[3]];
+          } else if (typeof token[3] === 'string' && hasDefault && !Array.isArray(route.defaults)) {
+            value = route.defaults[token[3]];
+          } else if (optional) {
+            return;
+          } else {
+            throw new Error('The route "' + name + '" requires the parameter "' + token[3] + '".');
+          }
+
+          let empty = true === value || false === value || '' === value;
+
+          if (!empty || !optional) {
+            let encodedValue = Router.encodePathComponent(value);
+
+            if ('null' === encodedValue && null === value) {
+              encodedValue = '';
+            }
+
+            url = token[1] + encodedValue + url;
+          }
+
+          optional = false;
+        } else if (hasDefault && (typeof token[3] === 'string' && token[3] in unusedParams)) {
+          delete unusedParams[token[3]];
+        }
+
+        return;
+      }
+
+      throw new Error('The token type "' + token[0] + '" is not supported.');
+    });
+
+    if (url === '') {
+      url = '/';
+    }
+
+    route.hosttokens.forEach((token) => {
+      let value;
+
+      if ('text' === token[0]) {
+        host = token[1] + host;
+
+        return;
+      }
+
+      if ('variable' === token[0]) {
+        if (token[3] in params) {
+          value = params[token[3]];
+          delete unusedParams[token[3]];
+        } else if (route.defaults && !Array.isArray(route.defaults) && (token[3] in route.defaults)) {
+          value = route.defaults[token[3]];
+        }
+
+        host = token[1] + value + host;
+      }
+    });
+    // Foo-bar!
```

Comment:
Should this be here?

Reply:
Good question, it was already there https://github.com/FriendsOfSymfony/FOSJsRoutingBundle/blob/master/Resources/js/router.js#L295 

But gonna remove it